### ISSUE TITLE
Consider creating easy-to-use library version/crate

### ISSUE THREAD

**Issue by Andrew15-5** [01mf02/jaq#188]
Hi. When I was searching for the Rust-native `jq` library (several months ago), I was surprised that I only found CLIs. But in a desperate attempt, I tried to "reverse-engineer" the CLI. I don't remember if jaq was the only viable option or not, but I chose it. It was extremely hard as there was almost no documentation for my case, so I started bisecting the CLI crate. In the end, the "boilerplate" became pretty huge. It definitely can be reduced if a dedicated library crate is used, but there is none.

<details><summary>the boilerplate</summary>

```rs
fn apply_json_filter<T>(json: serde_json::Value, filter: &str) -> Result<T>
where
    T: DeserializeOwned,
{
    let mut defs = ParseCtx::new(vec![]);
    defs.insert_natives(jaq_core::core());
    defs.insert_defs(jaq_std::std());

    let (main_filter, errs) = jaq_parse::parse(filter, jaq_parse::main());
    if !errs.is_empty() {
        return Err(make_error_message_from_vec(
            "There were some errors parsing the filter command:\n",
            errs,
            |e| format!("{}\n", e),
        ));
    }
    if main_filter.is_none() {
        return Err(Error::msg(
            "Main filter is `None` For some reason, even though there were no errors.",
        ));
    }

    let finished_filter = defs.compile(main_filter.unwrap());
    if !defs.errs.is_empty() {
        return Err(make_error_message_from_vec(
            "There were some errors compiling the main filter:\n",
            defs.errs,
            |e| format!("Error parsing characters in range: {:?}\n", e.1),
        ));
    }

    let inputs = RcIter::new(core::iter::empty());

    let out: Vec<_> = finished_filter
        .run((Ctx::new([], &inputs), Val::from(json)))
        .collect();

    if out.iter().any(|x| x.is_err()) {
        return Err(make_error_message_from_vec(
            "Failed to filter JSON object(s):\n",
            out.iter().filter(|x| x.is_err()).collect(),
            |e| format!("{:?}\n", e),
        ));
    }
    let objects: Vec<_> = out
        .into_iter()
        .map(|x| serde_json::Value::from(x.unwrap()))
        .collect();

    let filtered_json = if objects.len() == 1 {
        objects.into_iter().next().unwrap()
    } else {
        serde_json::Value::Array(objects)
    };

    serde_json::from_value(filtered_json).map_err(Error::from)
}
```

</details>

My implementation required at least 4 separate crates: `jaq-interpret`, `jaq-parse`, `jaq-core`, `jaq-std` (in addition to the `serde` and `serde_json`). I really hope that a library crate can be created that can provide good defaults in addition to some configuration (for optimizing performance and dependency tree). I also noticed that I'm not the first one who faced this problem and created an issue. Perhaps my function can be a starting point.

**Comment by 01mf02** [comment]
I agree with you. The current boilerplate is a bit too much for common use cases. I will think about how to make this better.

**Comment by 01mf02** [comment]
To understand your requirements better: Do you need precise parse error messages, or are you fine with getting some `Option<Filter>`? Are you using jaq for a fixed set of filters? Do you need to report errors to users?

**Comment by Andrew15-5** [comment]
I used my function for one specific type: `Vec<struct>`. But I think it would work for any other type just fine.

The filter is just a non-mutable raw string slice. It uses `select()` and `del()`. I think that in the future there can be a project where I need to use user input for a filter. Being able to choose a subset of things that can be used in a filter **if** it will reduce the file size (maybe slim down dependency tree) or increase performance would be a neat thing to have. By default, a `Default` (trait)/`Full` set should be used (either by providing some enum/type or having a separate function name for it).

Since there could be a few different errors when applying a filter, I would prefer having an option to distinguish between the errors. Because it will help me (as a developer) figure out where did I mess up. But some errors can be useful for the end user as well, so using `Result` would be nice. Plus, the `?` operator is a killer feature that a lot of people enjoy using. `Result` can always be converted to `Option`.

Perhaps some internal errors are not useful for the developer or the user. Such errors (in different places of the jaq's boilerplate) can be merged under the same error. But I used this library over a month ago (for a few days) so I don't remember any specifics about the errors. You know them better anyway.
